src/dbgshim/dbgshim.cpp

```diff
@@ -988,73 +1128,134 @@ GetTargetCLRMetrics(
     // Check the NT headers.
     if (!pedecoder.CheckNTFormat())
     {
-        ThrowHR(E_FAIL);
+        return E_FAIL;
     }
 
     // At this point we can safely read anything in the NT headers.
 
     if (!pedecoder.HasDirectoryEntry(IMAGE_DIRECTORY_ENTRY_EXPORT) ||
         !pedecoder.CheckDirectoryEntry(IMAGE_DIRECTORY_ENTRY_EXPORT))
     {
-        ThrowHR(E_FAIL);
+        return E_FAIL;
     }
-    IMAGE_DATA_DIRECTORY * pExportDirectoryEntry = pedecoder.GetDirectoryEntry(IMAGE_DIRECTORY_ENTRY_EXPORT);
-
-    // At this point we can safely read the IMAGE_DATA_DIRECTORY of the export directory.
 
-    if (!pedecoder.CheckDirectory(pExportDirectoryEntry))
+    // If we are looking for the DotNetRuntimeInfo export for a single-file app, do this before looking for
+    // engine metrics export ordinal for a faster out of the module search loop. There are plenty of other 
+    // native modules with the metrics ordinal #2.
+    if (pClrInfoOut != NULL)
     {
-        ThrowHR(E_FAIL);
+        if (IsCoreClr(wszModulePath))
+        {
+            PEDecoder_ResourceCallbackFunction callback = ([](LPCWSTR lpName, LPCWSTR lpType, DWORD langid, BYTE* data, COUNT_T cbData, void* context) { 
+                CLR_DEBUG_RESOURCE* pDebugResource = (CLR_DEBUG_RESOURCE*)data;
+                ClrInfo* pClrInfo = (ClrInfo*)context;
+                if (cbData != sizeof(CLR_DEBUG_RESOURCE) || pDebugResource->dwVersion != 0 || pDebugResource->signature != CLR_ID_ONECORE_CLR)
+                {
+                    return false;
+                }
+                pClrInfo->IndexType = LIBRARY_PROVIDER_INDEX_TYPE::Identity; 
+                pClrInfo->DbiTimeStamp = pDebugResource->dwDbiTimeStamp;
+                pClrInfo->DbiSizeOfImage = pDebugResource->dwDbiSizeOfImage;
+                pClrInfo->DacTimeStamp = pDebugResource->dwDacTimeStamp;
+                pClrInfo->DacSizeOfImage = pDebugResource->dwDacSizeOfImage;
+                return true;
+            });
+            if (!pedecoder.EnumerateWin32Resources(W("CLRDEBUGINFO"), MAKEINTRESOURCEW(10), callback, pClrInfoOut))
+            {
+                return E_FAIL;
+            }
+        }
+        else
+        { 
+            PTR_VOID runtimeInfoExport = pedecoder.GetExport(RUNTIME_INFO_SIGNATURE);
```

Comment:
Do you know how much this path added in terms of time? This worries me still as if a completely new failure point to go through all modules on attach, reading the exports. But as far as I can see it will only fail the attach and not cause any other major issues. So my only concern left is how much gets added to attach. If it's not something crazy, I think it's fine.

Reply:
If there are noticeable perf problems (so far in my simple testing there isn't), we can revisit this.  I don't think the only alternative of reading the module in memory with ReadProcessMemory (on Windows) would be that much faster.  

This PR doesn't fix Windows AppX bundles which will probably require us to switch to ReadProcessMemory and come up with a PE reader that can do "loaded" layout (PEDecoder only does file layout currently) remotely.